File: solutions/javascript/lasagna-master/1/lasagna-master.js
/// <reference path="./global.d.ts" />
// @ts-check

/**
 * Implement the functions needed to solve the exercise here.
 * Do not forget to export them so they are available for the
 * tests. Here an example of the syntax as reminder:
 *
 * export function yourFunction(...) {
 *   ...
 * }
 */
export function cookingStatus(remainingTime) {
    switch (remainingTime) {
        case 0:
            return 'Lasagna is done.';
        case undefined:
            return 'You forgot to set the timer.';
        default:
            return 'Not done, please wait.';
    }
}

export function preparationTime(layers, time) {
    return layers.length * (time ?? 2);
}

export function quantities(layers) {
    let noodleCount = 0;
    let sauceCount = 0;

    for (let i = 0; i < layers.length; i++) {
        if (layers[i] === 'noodles') {
            noodleCount++;
        } else if (layers[i] === 'sauce') {
            sauceCount++;
        }
    }

    return {
        'noodles': noodleCount * 50,
        'sauce': sauceCount * 0.2
    };
}

export function addSecretIngredient(friendsList, myList) {
    myList.push(friendsList[friendsList.length - 1]);
}

export function scaleRecipe(recipe, portions) {
    let newRecipe = structuredClone(recipe);
    let scale = portions / 2;

    for (let key in newRecipe) {
        newRecipe[key] *= scale;
    }

    return newRecipe;
}

File: solutions/go/lasagna-master/1/lasagna_master.go
// Package lasagna provides functions for managing lasagna recipes.
package lasagna

// PreparationTime returns the time to prepare a lasagna,
// given the layers and the time to cook each one.
func PreparationTime(layers []string, time int) int {
	if time <= 0 {
		time = 2
	}

	return len(layers) * time
}

// Quantities returns the grams of noodles and sauce needed for a lasagna,
// given the layers of the lasagna.
func Quantities(layers []string) (int, float64) {
	var gramsNoodles int = 0
	var gramsSauce float64 = 0

	for _, v := range layers {
		if v == "noodles" {
			gramsNoodles += 50
		} else if v == "sauce" {
			gramsSauce += 0.2
		}
	}

	return gramsNoodles, gramsSauce
}

// AddSecretIngredient modifies a recipe for a given lasagna,
// given another recipe with the secret ingredient at the end.
func AddSecretIngredient(friendsList []string, myList []string) {
	friendListLength := len(friendsList)
	myListLength := len(myList)

	myList[myListLength-1] = friendsList[friendListLength-1]
}

// ScaleRecipe returns the scaled up quantities,
// given the initial quantities and poritons.
func ScaleRecipe(quantities []float64, portions int) []float64 {
	var newQuantities []float64

	for _, v := range quantities {
		newQuantities = append(newQuantities, float64(portions)/2*v)
	}

	return newQuantities
}
